### PR TITLE
Refactor js messages component to observables

### LINKED ISSUE #28692: CustomerData messages do not render
### Preconditions (*)

1. Magento 2.3.5-p1

### Steps to reproduce (*)

1. Trigger error message here: `\Magento\CatalogInventory\Model\StockStateProvider::checkQuoteItemQty()` "The fewest you may purchase is %1" by adding something to cart that is less than the minimum product is configured for. For example, a bundled product, with a child product that has "Option Min Qty" of 2 but only entering 1. (Note: it may be necessary to disable Javascript, change the input, and then submit).
2. Watch for the `customer/section/load` response with a `messages` key in the object.

### Expected result (*)

1. A message should show near the header (via `messages.phtml` in the `messages()` area)
2. The `this.messages` should not be so tightly coupled to `customerData` that is clears `this.messages` when it clears `customerData`'s messages (see: `vendor/magento/module-theme/view/frontend/web/js/view/messages.js`)

### Actual result (*)

1. `this.messages` in `vendor/magento/module-theme/view/frontend/web/js/view/messages.js` is cleared with `customerData.set('messages', {});`

More details based on how this runs `vendor/magento/module-theme/view/frontend/web/js/view/messages.js`:
1. `this.messages` default is an empty array – even though it would crash the Knockout if initialized with.
2. `this.initialize()` runs before the Knockout is initialized and `this.messages` is set from `customerData`.
3. If `this.messages().messages` is not empty (and it won't be if there was a message returned), the `customerData` version of the messages is cleared.
5. *Problem:* that `customerData` observable is the same one the messages one references. As a result, clearing it clears the messages.

One possible solution would involve giving messages it's own observable and getting the data from the `customerData` module.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

giving messages it's own ability array observable track by K.O and getting the data from the `customerData` messages
main purpose to Isolates from customerData (Default magento store messages in this object). Manage message in messages observables property only

Current magento get messages via `$.cookieStorage.get('mage-messages')` instead of internal `messages`. The first `if` block
default working for show messages. Logic here check messages from `cookieMessages`

Screenshot from 2021-03-26 09-12-53
The second `if` block may never have a chance to do some render message. the variables messages may work but in another case it's may be cleared and not display message

the messages are retrieved from customerData. And next line is removing the customerData messages. This will also remove the values from the variable this.messages which is later on called in template.
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Theme/view/frontend/web/js/view/messages.js#L39

### Fixed Issues (if relevant)

1. Fixes magento/magento2#28692
Seem has closed
### Manual testing scenarios (*)

1. ...
2. ...

### Questions or comments

CC: @bassplayer7 maybe you can give me some reviews
@omiroshnichenko @nathanjosiah @zakdma
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by srsathish92** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Unit Tests, Database Compare, Functional Tests CE, Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

File: app/code/Magento/Theme/view/frontend/web/js/view/messages.js [hunk 1]
```diff
@@ -8,32 +8,48 @@
...
     'jquery/jquery-storageapi'
-], function ($, Component, customerData, _, escaper) {
+], function ($, ko, Component, customerData, _, escaper) {
     'use strict';
 
     return Component.extend({
         defaults: {
             cookieMessages: [],
-            messages: [],
+            messages: ko.observableArray([]),
```
Comment (omiroshnichenko) [inline app/code/Magento/Theme/view/frontend/web/js/view/messages.js]: This is a backward incompatible change. If you want to make it "observable", use `tracks` instead. More information you could find [here](https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/ui_comp_uielement_concept.html)

**Review by omiroshnichenko — CHANGES_REQUESTED** [review]

File: app/code/Magento/Theme/view/frontend/web/js/view/messages.js (hunk 1 above)
Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/web/js/view/messages.js]: I have added track variable in initObservable

Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/web/js/view/messages.js]: Current when digging here i found that magento not pass message via this `messages`. It's empty may never used.
Magento get messages from `cookieStorage.get('mage-messages')`. Store error messages in local browser of customer

Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/web/js/view/messages.js]: https://github.com/js-cookie/js-cookie
https://github.com/julien-maurel/js-storage
Maybe this one will be use in near future

What we current used https://github.com/julien-maurel/jQuery-Storage-API (seem no longer maintain). I'm not sure any workaround internal related this

**Comment by gabrieldagama** [comment]
Hi @mrtuvn,

Can you provide a detailed explanation on how this issue can impact the customer? Please also add some steps to reproduce, the ones from the issue itself aren't very clear.

Thanks.

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Pull request closed without merging**